### PR TITLE
Miscellneous refactorings of trans

### PR DESCRIPTION
**Pull request by nikomatsakis**
This doesn't achieve any particular goal yet, but it's a collection of refactorings with the common goal of turning `SharedCrateContext` etc into stuff that we can use with on-demand and actually expect to hash in a stable fashion for incremental. Not there yet, clearly. 

r? @eddyb 
cc @michaelwoerister

### CODE REVIEW

File: src/librustc/middle/cstore.rs [hunk 1]
```diff
@@ -52,7 +52,6 @@ pub use self::NativeLibraryKind::*;
 
 #[derive(Clone, Debug)]
 pub struct LinkMeta {
-    pub crate_name: Symbol,
     pub crate_hash: Svh,
 }
```
Comment (eddyb) [inline src/librustc/middle/cstore.rs]: This is one of the weirder structures in the whole compiler and it'd be nice if it'd just die already.

File: src/librustc_trans/base.rs [hunk 2]
```diff
@@ -784,15 +791,16 @@ fn write_metadata(cx: &SharedCrateContext,
         // metadata doesn't get loaded into memory.
         let directive = format!(".section {}", section_name);
         let directive = CString::new(directive).unwrap();
-        llvm::LLVMSetModuleInlineAsm(cx.metadata_llmod(), directive.as_ptr())
+        llvm::LLVMSetModuleInlineAsm(metadata_llmod, directive.as_ptr())
     }
-    return metadata;
+    return (metadata_llcx, metadata_llmod, metadata);
```
Comment (eddyb) [inline src/librustc_trans/base.rs]: I am surprised we don't have a wrapper for the `llcx` + `llmod` combo.

File: src/librustc_trans/base.rs [hunk 3]
```diff
@@ -1055,28 +1063,28 @@ pub fn trans_crate<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
...
-    });
+    let (metadata_llcx, metadata_llmod, metadata) =
+        time(tcx.sess.time_passes(), "write metadata", || {
+            write_metadata(tcx, &link_meta, shared_ccx.exported_symbols())
+        });
 
     let metadata_module = ModuleTranslation {
         name: link::METADATA_MODULE_NAME.to_string(),
         symbol_name_hash: 0, // we always rebuild metadata, at least for now
         source: ModuleSource::Translated(ModuleLlvm {
```
Comment (eddyb) [inline src/librustc_trans/base.rs]: `ModuleLlvm` is what I mean I guess?

File: src/librustc_trans/base.rs [hunk 4]
```diff
@@ -1191,26 +1205,38 @@ pub fn trans_crate<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
             if ccx.sess().opts.debuginfo != NoDebugInfo {
                 debuginfo::finalize(&ccx);
             }
-        }
+
+            ModuleTranslation {
+                name: cgu_name,
+                symbol_name_hash,
+                source: ModuleSource::Translated(ModuleLlvm {
```
Comment (eddyb) [inline src/librustc_trans/base.rs]: Again, it'd be nice if `ccx` actually kept a `ModuleLlvm`.

File: src/librustc_trans/context.rs [hunk 5]
```diff
@@ -59,19 +58,30 @@ pub struct Stats {
     pub fn_stats: RefCell<Vec<(String, usize)> >,
 }
 
+impl Stats {
+    pub fn extend(&mut self, stats: Stats) {
+        self.n_glues_created.set(self.n_glues_created.get() + stats.n_glues_created.get());
+        self.n_null_glues.set(self.n_null_glues.get() + stats.n_null_glues.get());
+        self.n_real_glues.set(self.n_real_glues.get() + stats.n_real_glues.get());
```
Comment (eddyb) [inline src/librustc_trans/context.rs]: There's a bunch of stats that are irrelevant anymore AFAIK.

**Review by eddyb — APPROVED** [review]
LGTM modulo some nits that can be addressed later.

File: src/librustc/middle/cstore.rs (hunk 1 above)
Comment (nikomatsakis) [inline src/librustc/middle/cstore.rs]: oh yeah I meant to kill it altogether; it's just a newtype'd svh at this point

Comment (nikomatsakis) [inline src/librustc/middle/cstore.rs]: I can do that in a follow-up though

**Pull request merged**